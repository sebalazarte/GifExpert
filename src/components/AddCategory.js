import React, { useState, Fragment } from 'react'
import PropTypes from 'prop-types'


export const AddCategory = ({ setCategorias, setLimite }) => { //se desestructuro props

    const [inputValue, setInputValue] = useState('');
    const [numberValue, setNumberValue] = useState(10);

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleNumberChange = (e) => {
        setNumberValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategorias(cats => [inputValue, ...cats,]);
            setInputValue('');
        }

        if (numberValue.length > 0) {
            setLimite(numberValue);
        }
    }

    const handleClean = (e) => {
        e.preventDefault();
        setCategorias([]);
    }

    return (
        <Fragment>
            <form onSubmit={handleSubmit}>
                <div className="row g-3">
                    <div className="col-md-8">
                        <div className="input-group mb-3">
                            <label className="input-group-text">Buscar</label>
                            <input
                                id="txtBuscar"
                                type="text"
                                value={inputValue}
                                onChange={handleInputChange}
                                className="form-control"
                                placeholder="ingrese para buscar..."
                                required />
                            <label className="input-group-text">Limite</label>
                            <input
                                id="txtCantidad"
                                type="number"
                                value={numberValue}
                                onChange={handleNumberChange}
                                className="form-control cantidad"
                            />
                            <button type="submit" className="btn btn-primary">
                                Buscar
                            </button>
                        </div>
                    </div>
                    <div className="col-md-4">
                        <button onClick={handleClean} className="btn btn-danger">
                            Limpiar
                        </button>   
                    </div>
                </div>
            </form>

        </Fragment>
    )
}

AddCategory.propTypes = {
    setCategorias: PropTypes.func.isRequired
}