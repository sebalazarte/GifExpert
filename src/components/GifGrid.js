import React, {Fragment } from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';


export const GifGrid = ({ category, limit }) => {

    const {data: images, loading} = useFetchGifs(category, limit);

    return (
        <Fragment>
            <h3 className="animate__animated animate__faden">{category}</h3>
            
            {loading && <p className="animate__animated animate__flash">Cargando...</p>}

            <div className="card-grid">

                {
                    images.map(img =>
                        <GifGridItem
                            key={img.id}
                            {...img} />)
                }
            </div>
        </Fragment>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
    limit: PropTypes.number.isRequired
}