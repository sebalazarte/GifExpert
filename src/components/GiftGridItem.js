import React from 'react'

export const GiftGridItem = ({id, title, url}) => {
    // console.log(props);
    return (
        <div className="card animate__animated animate__bounceIn">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}
