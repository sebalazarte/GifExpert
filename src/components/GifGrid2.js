import React, { useState, useEffect, Fragment } from 'react'
import { getGifts } from '../helpers/getGifs';
import { GiftGridItem } from './GiftGridItem';

export const GifGrid = ({ category }) => {

    const [images, setImages] = useState([]);

    useEffect(()=>{
        getGifts(category)
            .then( setImages );
    }, [category]);

    return (
        <Fragment>
            <h3>{category}</h3>
            <div className="card-grid">

                {
                    images.map(img =>
                        <GiftGridItem
                            key={img.id}
                            {...img} />)
                }
            </div>
        </Fragment>
    )
}
