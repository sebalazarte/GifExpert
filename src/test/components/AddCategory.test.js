import React from 'react';
import { shallow } from 'enzyme';
import { AddCategory } from '../../components/AddCategory';
import '@testing-library/jest-dom';

describe('Pruebas en el componente <AddCategory/>', () => {

    const setCategorias = jest.fn();
    const setLimite = jest.fn();

    let wraper = shallow(
        <AddCategory setCategorias={setCategorias} setLimite={setLimite} />);


    beforeEach(() => {
        jest.clearAllMocks();
        wraper = shallow(
            <AddCategory setCategorias={setCategorias} setLimite={setLimite} />);
    });

    test('Debe mostrarse correctamente', () => {
        expect(wraper).toMatchSnapshot();
    });

    test('Debe de cambiar el input[text]', () => {
        const input = wraper.find('#txtBuscar');
        const value = 'hola mundo';
        input.simulate('change', {
            target: { value }
        });

    });

    test('No debe llamar a submit si el Input text no tiene valor', () => {
        wraper.find('form').simulate('submit', { preventDefault() { } });
        expect(setCategorias).not.toHaveBeenCalled();
    });

    test('Debe de llamar SetCategorias y limpiar el Input[Text] ', () => {
        
        //1. simular el input change
        const input = wraper.find('#txtBuscar');
        const value = 'hola mundo';
        input.simulate('change', {
            target: { value }
        });

        //2. simular el submit
        wraper.find('form').simulate('submit', { preventDefault() { } });
        
        //3.SetCategorias se debe haber llamado
        expect(setCategorias).toHaveBeenCalled();
        expect(setCategorias).toHaveBeenCalledTimes(1);
        expect(setCategorias).toHaveBeenCalledWith(expect.any(Function));

        //4. El valor el input debe estar vacio
        expect(input.prop('value')).toBe('');
    });

});