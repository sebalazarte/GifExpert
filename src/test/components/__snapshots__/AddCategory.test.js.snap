// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <AddCategory/> Debe mostrarse correctamente 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <div
      className="row g-3"
    >
      <div
        className="col-md-8"
      >
        <div
          className="input-group mb-3"
        >
          <label
            className="input-group-text"
          >
            Buscar
          </label>
          <input
            className="form-control"
            id="txtBuscar"
            onChange={[Function]}
            placeholder="ingrese para buscar..."
            required={true}
            type="text"
            value=""
          />
          <label
            className="input-group-text"
          >
            Limite
          </label>
          <input
            className="form-control cantidad"
            id="txtCantidad"
            onChange={[Function]}
            type="number"
            value={10}
          />
          <button
            className="btn btn-primary"
            type="submit"
          >
            Buscar
          </button>
        </div>
      </div>
      <div
        className="col-md-4"
      >
        <button
          className="btn btn-danger"
          onClick={[Function]}
        >
          Limpiar
        </button>
      </div>
    </div>
  </form>
</Fragment>
`;
